cpp/src/arrow/acero/source_node.cc

```diff
@@ -263,6 +263,22 @@ struct SourceNode : ExecNode, public TracedNode {
     to_finish.MarkFinished();
   }
 
+  Status StopProducing() override {
+    // GH-35837: ensure node is not paused
+    Future<> to_finish;
+    {
+      std::lock_guard<std::mutex> lg(mutex_);
+      if (!backpressure_future_.is_finished()) {
+        to_finish = backpressure_future_;
```

Comment:
```suggestion
        to_finish = backpressure_future_;
        backpressure_future_ = Future<>::MakeFinished();
```

We might want to do this here and in `ResumeProducing`.  Otherwise, if `ResumeProducing` and `StopProducing` are called at the same time then both calls might try and mark the same future finished (which isn't valid).

Reply:
@rtpsw Is this addressed?

Reply:
Yes, [here](https://github.com/apache/arrow/pull/35902/files#diff-0f58fd9023e0533b4e9149839729afa86c16341c1bd5ddd8c7578ef6bc441968R262) and [here](https://github.com/apache/arrow/pull/35902/files#diff-0f58fd9023e0533b4e9149839729afa86c16341c1bd5ddd8c7578ef6bc441968R274).